ci/run_tests_test.go

```diff
@@ -60,3 +65,144 @@ func TestRunTests(t *testing.T) {
 	}
 	t.Logf("\n%s\nExecTime: %v\nSecret: %v\n", ed.out, ed.execTime, info.RandomSecret)
 }
+
+func TestRecordResults(t *testing.T) {
+	f, err := ioutil.TempFile(os.TempDir(), "testing")
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := f.Close(); err != nil {
+		os.Remove(f.Name())
+		t.Fatal(err)
+	}
+
+	db, err := database.NewGormDB(f.Name(), zap.NewNop())
+	if err != nil {
+		os.Remove(f.Name())
+		t.Fatal(err)
+	}
+
+	// clean up the temp database file
+	defer func() {
+		if err := os.Remove(f.Name()); err != nil {
+			t.Error(err)
+		}
+	}()
+
+	var user pb.User
+	if err := db.CreateUserFromRemoteIdentity(&user,
+		&pb.RemoteIdentity{
+			Provider:    "fake",
+			RemoteID:    1,
+			AccessToken: "token",
+		}); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := db.UpdateUser(&pb.User{ID: user.ID, IsAdmin: true}); err != nil {
+		t.Fatal(err)
+	}
+
+	course := &pb.Course{
+		Name:            "Test",
+		OrganizationID:  1,
+		CourseCreatorID: user.ID,
+		SlipDays:        5,
+	}
+	if err := db.CreateCourse(user.ID, course); err != nil {
+		t.Fatal(err)
+	}
+
+	assignment := &pb.Assignment{
+		CourseID:         course.ID,
+		Name:             "lab1",
+		ScriptFile:       "go.sh",
+		Deadline:         "2022-11-11T13:00:00",
+		AutoApprove:      true,
+		ScoreLimit:       70,
+		Order:            1,
+		IsGroupLab:       false,
+		ContainerTimeout: 1,
+	}
+	if err = db.CreateAssignment(assignment); err != nil {
+		t.Fatal(err)
+	}
+
+	buildInfo := &score.BuildInfo{
+		BuildDate: "2022-11-10T13:00:00",
+		BuildLog:  "Testing",
+		ExecTime:  33333,
+	}
+	testScores := []*score.Score{
+		{
+			Secret:   "secret",
+			TestName: "Test",
+			Score:    10,
+			MaxScore: 15,
+			Weight:   1,
+		},
+	}
+
+	// Must create a new submission with correct scores and build info, not approved
+	results := &score.Results{
+		BuildInfo: buildInfo,
+		Scores:    testScores,
+	}
+	runData := &RunData{
+		Course:     course,
+		Assignment: assignment,
+		Repo: &pb.Repository{
+			UserID: 1,
+		},
+		JobOwner: "test",
+	}
+
+	recordResults(zap.NewNop().Sugar(), db, runData, results)
+
+	submission, err := db.GetSubmission(&pb.Submission{AssignmentID: assignment.ID, UserID: user.ID})
+	if err != nil {
+		t.Fatal(err)
+	}
+	if submission.Status == pb.Submission_APPROVED {
+		t.Fatal("Submission must not be auto approved")
+	}
+	if !reflect.DeepEqual(testScores, submission.Scores) {
```

Comment:
We should use `cmp.Diff` instead of `reflect.DeepEqual` for all new code. See examples elsewhere in the codebase. PS: Use the `protocmp.Transform()` approach, if the input arguments are proto messages.

Reply:
Here is one example:
```golang
if diff := cmp.Diff(wantAssignment, assignments[0], protocmp.Transform()); diff != "" {
	t.Errorf("GetAssignmentsWithSubmissions() mismatch (-want +got):\n%s", diff)
}
```

Reply:
Changed.